src/test/fuzz/simplicity_tx.cpp

```diff
@@ -0,0 +1,236 @@
+                    return;
+                }
+            }
+        }
+
+        // We use the first vin as a "random oracle" rather than reading more from
+        // the fuzzer, because we want our fuzz seeds to have as simple a structure
+        // as possible. This means we must reject 0-input transactions, which are
+        // invalid on-chain anyway.
+        if (mtx.vin.size() == 0) {
+            return;
+        }
+    }
+    const auto& random_bytes = mtx.vin[0].prevout.hash;
+
+    // 3. Construct `nIn` and `spent_outs` arrays.
+    bool expect_simplicity = false;
+    std::vector<unsigned char[32]> cmrs;
+    std::vector<CTxOut> spent_outs{};
+    for (unsigned int i = 0; i < mtx.vin.size(); i++) {
+        // Null asset or value would assert in the interpreter, and are impossible
+        // to hit in real transactions. Nonces are not included in the UTXO set and
+        // therefore don't matter.
+        CConfidentialValue value = i & 1 ? INPUT_VALUE_CONF : INPUT_VALUE_UNCONF;
+        CConfidentialAsset asset = i & 2 ? INPUT_ASSET_CONF : INPUT_ASSET_UNCONF;
+        CScript scriptPubKey;
+        if (i < random_bytes.size()) {
+            if (i & 1 && random_bytes.data()[i] & 1) {
+                value.vchCommitment[0] ^= 1;
+            }
+            if (i & 2 && random_bytes.data()[i] & 2) {
+                asset.vchCommitment[0] ^= 1;
+            }
+        }
+
+        // Check for size 4: a Simplicity program will always have a witness, program,
+        // CMR, control block and (maybe) annex, in that order. If the annex is present,
+        // then checking for size 4 doesn't guarantee that a witness is present, but
+        // that is ok at this point. (In fact, it is a useful thing to check.)
+        if (i < mtx.witness.vtxinwit.size()) {
+            auto& current = mtx.witness.vtxinwit[i].scriptWitness.stack;
+            if (current.size() >= 4) {
+                size_t top = current.size();
+                if (!current[top - 1].empty() && current[top - 1][0] == 0x50) {
+                    --top;
+                }
+                const auto& control = current[top - 1];
+                const auto& program = current[top - 3];
+
+                if (control.size() >= TAPROOT_CONTROL_BASE_SIZE && (control[0] & 0xfe) == 0xbe) {
+                    // The fuzzer won't be able to produce a valid CMR on its own, so we compute it
+                    // and jam it into the witness stack. But we do require the fuzzer give us a
+                    // place to put it, so we don't have to resize the stack (and so that actual
+                    // valid transactions will work with this code).
+                    // Compute CMR and do some sanity checks on it (and the program)
+                    std::vector<unsigned char> cmr(32, 0);
+                    assert(cmr.size() == 32); // fuck C++
```

Comment:
In 09de430f956c32af6d010bd52b7a11e9b9e5e51e: lol should probably remove this too.